### ISSUE TITLE
defaultErrorHandler causes inappropriate 406 reponses

### ISSUE THREAD

**Issue by whittle** [yesodweb/yesod#1402]
Copying over from a thread at https://groups.google.com/forum/#!topic/yesodweb/ed6MvRG_pVg

I’m seeing some odd behavior with a handler that uses `selectRep` to choose a representation. I’m on LTS-8.5. The handler in question looks like this: 

```
getVideoIconR :: VideoId -> Handler TypedContent
getVideoIconR i = do
  a <- runDB $ get404 i
  t <- searchForImageUrl a
  selectRep $ do
    provideRepType "image/*" $ (redirect t :: Handler ())
    provideJson $ object ["url" .= t]
```

I have specs covering four cases (excluding cases that test the behavior of `searchForImageUrl`): 

1. Request accepting "application/json" with a VideoId that exists in the database. _Responds with a 200 and a body containing the expected JSON._ 
2. Request accepting "image/*" with a VideoId that exists in the database. _Responds with a 303 and a Location header with the expected URL._ 
3. Request accepting "application/json" with a VideoId that does not exist in the database. _Responds with a 404._ 
4. Request accepting "image/*" with a VideoId that does not exist in the database. _Responds with a 406._ 

The last test case is not what I would expect; I would have thought that short-circuiting execution of the handler with `notFound` would prevent the server from attempting to find a representation for data that won’t be returned anyway. Failing that, I would expect both the representations to behave similarly. 

It looks like defaultErrorHandler uses selectRep to choose a representation for >=400 responses, which can generate an inappropriate 406 response when the error handler doesn’t explicitly provide an acceptable body. This theory is bolstered by replacing `defaultErrorHandler` in my app with a custom error handler that explicitly handles acceptance of "image/*" causes the app to respond with a 404 in the fourth case I listed previously.

**Comment by snoyberg** [comment]
Can you follow up on this with a PR? Ideal would be if you included a fix for this, but even just PRing the failing test case would be great.

**Comment by StevenXL** [comment]
Closing as this was addressed by #1540.
